### PR TITLE
rpc: implement full bi-directional communication

### PR DESCRIPTION
**Pull request by fjl**
New APIs added:

    client.RegisterName(namespace, service) // makes service available to server
    client.Notify(ctx, method, args...)     // sends a notification
    ClientFromContext(ctx)                  // to get a client in handler method

This is essentially a rewrite of the server-side code. JSON-RPC
processing code is now the same on both server and client side. Many
minor issues were fixed in the process and there is a new test suite for
JSON-RPC spec compliance (and non-compliance in some cases).

List of behavior changes:

- Method handlers are now called with a per-request context instead of a
  per-connection context. The context is canceled right after the method
  returns.
- Subscription error channels are always closed when the connection
  ends. There is no need to also wait on the Notifier's Closed channel
  to detect whether the subscription has ended.
- Parse errors just close the connection instead of reporting an error
  back to the client, except on HTTP connections. I'm not happy about
  this change, but it was non-trivial to implement this within the
  concurrency model used in Client.
- Client now omits "params" instead of sending "params": null when there
  are no arguments to a call. The previous behavior was not compliant
  with the spec. The server still accepts "params": null.
- Floating point numbers are allowed as "id". The spec doesn't allow
  them, but we handle request "id" as json.RawMessage and guarantee that
  the same number will be sent back.
- Logging is improved significantly. There is now a message at DEBUG
  level for each RPC call served.

Supersedes #18075 (this PR implements https://github.com/ethereum/go-ethereum/pull/18075#issuecomment-438291381)
Supersedes #18078 (suitableCallbacks is much simplifed)

### CODE REVIEW

File: rpc/json.go [hunk 1]
```diff
@@ -18,71 +18,111 @@ package rpc
...
 	"strings"
 	"sync"
-
-	"github.com/ethereum/go-ethereum/log"
+	"time"
 )
 
 const (
-	jsonrpcVersion           = "2.0"
+	vsn                      = "2.0"
```
Comment (holiman) [inline rpc/json.go]: Three letter abbreviation of `version`, really? Please use `version` or leave it as it was previously

Comment (fjl) [reply rpc/json.go]: IMHO the constant name has to be shorter than `"2.0"`, otherwise it's not worth defining it.

**Comment by fjl** [comment]
Looks like the last commit fixed the websocket issue.

File: rpc/json.go (hunk 1 above)
Comment (fjl) [reply rpc/json.go]: I will change it to `version`.

**Comment by holiman** [comment]
Finally support for `pi` in `id`
```
curl -X POST --data '{"jsonrpc":"2.0","method":"web3_clientVersion","params":[],"id":3.14159}' -H "Content-type:application/json" http://localhost:8545
{"jsonrpc":"2.0","id":3.14159,"result":"Geth/v1.8.22-unstable-eddff1b2/linux-amd64/go1.11.2"}
```

Also, regarding parse errors, this still seems to work:
```
> eth.getUncle("0xasdfasdfasdfasdfasdfasdfadsasdfasdfasdfasdfasdfadsfasdffasdfasdf")
Error: invalid argument 0: json: cannot unmarshal invalid hex string into Go value of type common.Hash
```
But basically, these break (master below)
```
[user@work go-ethereum]$ echo '{"jsonrpc":"2.0","method":"web3_clientVersion,"params":[],"id":3.14}' | nc  -U  ~/.ethereum/geth.ipc
{"jsonrpc":"2.0","error":{"code":-32600,"message":"invalid character 'p' after object key:value pair"}}
[user@work go-ethereum]$ echo '{BAZONK "jsonrpc":"2.0","method":"web3_clientVersion,"params":[],"id":3.14}' | nc  -U  ~/.ethereum/geth.ipc
{"jsonrpc":"2.0","error":{"code":-32600,"message":"invalid character 'B' looking for beginning of object key string"}}
```
This PR: 
```
$ echo '{"jsonrpc":"2.0","method":"web3_clientVersion,"params":[],"id":3.14}' | nc  -U  ~/.ethereum/geth.ipc
$ echo '{"jsonrpc":"2.0","method":"web3_clientVersion,"params":[],"id":3.14}' | nc  -U  ~/.ethereum/geth.ipc
```
I'd reckon that most errors are not on the `json` format level, but rather on the layer above that, so I think that's ok, if not ideal.

**Comment by holiman** [comment]
This is odd. The first one is against `master`, the second against this PR: 
```
[user@work go-ethereum]$ echo '{"jsonrpc":"2.0","method":"web3_clientVersion","params":[""],"id":3.14}' | nc  -U  ~/.ethereum/geth.ipc
{"jsonrpc":"2.0","id":3.14,"result":"Geth/v1.9.0-unstable-ae857e74/linux-amd64/go1.11.2"}
[user@work go-ethereum]$ echo '{"jsonrpc":"2.0","method":"web3_clientVersion","params":[""],"id":3.14}' | nc  -U  ~/.ethereum/geth.ipc
```
It doesn't allow `[""]`. But after removing that: 
```
[user@work go-ethereum]$ echo '{"jsonrpc":"2.0","method":"web3_clientVersion","params":[],"id":3.14}' | nc  -U  ~/.ethereum/geth.ipc
[user@work go-ethereum]$ 
```
There's still no response?

**Comment by holiman** [comment]
Ah, there's something racey here.. 
```
[user@work go-ethereum]$ echo -n '{"jsonrpc":"2.0","method":"web3_clientVersion","id":4.14}' | nc  -U  ~/.ethereum/geth.ipc
[user@work go-ethereum]$ echo -n '{"jsonrpc":"2.0","method":"web3_clientVersion","id":4.14}' | nc  -U  ~/.ethereum/geth.ipc
[user@work go-ethereum]$ echo -n '{"jsonrpc":"2.0","method":"web3_clientVersion","id":4.14}' | nc  -U  ~/.ethereum/geth.ipc
[user@work go-ethereum]$ echo -n '{"jsonrpc":"2.0","method":"web3_clientVersion","id":4.14}' | nc  -U  ~/.ethereum/geth.ipc
[user@work go-ethereum]$ echo -n '{"jsonrpc":"2.0","method":"web3_clientVersion","id":4.14}' | nc  -U  ~/.ethereum/geth.ipc
{"jsonrpc":"2.0","id":4.14,"result":"Geth/v1.8.22-unstable-eddff1b2/linux-amd64/go1.11.2"}

```

**Comment by holiman** [comment]
Could it be that the `drain` triggers the `EOF`, which is sent over teh `errc` channel, and that happens simultaneously as the request is being processed?

**Comment by holiman** [comment]
One from `inproc.go`, and another from `server.go`. They seem to be competing

**Comment by holiman** [comment]
Hm, the one from `inproc` comes from here: https://github.com/ethereum/go-ethereum/blob/master/cmd/geth/main.go#L303

**Comment by holiman** [comment]
Oddly, even if I remove that whole chunk of self-attach code there's still some race. The only real difference is that I don't notice it until later, when the request comes in (if I leave the chunk in, it becomes triggered earlier) 

However, it appears that ServeCodec somehow spins up two cliens, or alternatively invokes the dispatcher twice, not quite sure.

**Comment by holiman** [comment]
Ok, I think I found it. In `client.go`, there's a loop that reads data. 
```
func (c *Client) read(codec ServerCodec) {
	for {
		log.Info("client starting read", "client.isHTTP", c.isHTTP, "client", c)
		msgs, batch, err := codec.Read()
		log.Info("client read data", "msgs", len(msgs), "err", err)
		if err != nil {
			c.readErr <- err
			return
		}
		c.readOp <- readOp{msgs, batch}
		log.Info("client delivered data", "batch", batch)
	}
}
```
yields
```
INFO [01-23|20:44:17.823] client starting read                     client.isHTTP=false client="&{idgen:0x858ce0 isHTTP:false services:0xc004723980 idCounter:0 reconnectFunc:<nil> writeConn:0xc004bf3f90 close:0xc005c3e2a0 closing:0xc005c3e300 didClose:0xc005c3e360 reconnected:0xc005c3e3c0 readOp:0xc005c3e420 readErr:0xc005c3e480 reqInit:0xc005c3e4e0 reqSent:0xc005c4e600 reqTimeout:0xc005c3e540}"
INFO [01-23|20:44:17.828] client read data                         msgs=1 err=nil
INFO [01-23|20:44:17.829] client delivered data                    batch=false
INFO [01-23|20:44:17.829] client starting read                     client.isHTTP=false client="&{idgen:0x858ce0 isHTTP:false services:0xc004723980 idCounter:0 reconnectFunc:<nil> writeConn:0xc004bf3f90 close:0xc005c3e2a0 closing:0xc005c3e300 didClose:0xc005c3e360 reconnected:0xc005c3e3c0 readOp:0xc005c3e420 readErr:0xc005c3e480 reqInit:0xc005c3e4e0 reqSent:0xc005c4e600 reqTimeout:0xc005c3e540}"
INFO [01-23|20:44:17.829] client read data                         msgs=0 err=EOF
```
So the `client` first reads the json data, and delivers over `c.readOp`. Then it immediately reads again, but now the remote side is finished, and there's an `EOF`. The eof is passed via `c.readErr <- err`, which immediately closes the conn. 

Not sure about the correct way to fix that...

**Comment by holiman** [comment]
Possible fix (though not complete)
```diff
diff --git a/rpc/client.go b/rpc/client.go
index 6ca5b3c6c..9fc327a94 100644
--- a/rpc/client.go
+++ b/rpc/client.go
@@ -22,6 +22,7 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"io"
 	"net/url"
 	"reflect"
 	"strconv"
@@ -533,10 +534,9 @@ func (c *Client) dispatch(codec ServerCodec) {
 		}
 		close(c.didClose)
 	}()
-
 	// Spawn the initial read loop.
 	go c.read(codec)
-
+	responseDoneCh := make(chan (bool), 1)
 	for {
 		select {
 		case <-c.close:
@@ -547,10 +547,16 @@ func (c *Client) dispatch(codec ServerCodec) {
 			if op.batch {
 				conn.handler.handleBatch(op.msgs)
 			} else {
-				conn.handler.handleMsg(op.msgs[0])
+				conn.handler.handleMsg(op.msgs[0], responseDoneCh)
 			}
 
 		case err := <-c.readErr:
+			if err == io.EOF {
+				// No more reading, but might not want to close it immediately, in case
+				// we are still processing and wants to write data.
+				<- responseDoneCh
+
+			}
 			conn.handler.log.Debug("RPC connection read error", "err", err)
 			conn.close(err, lastOp)
 			reading = false
diff --git a/rpc/handler.go b/rpc/handler.go
index bc03ef25f..44a6c2d73 100644
--- a/rpc/handler.go
+++ b/rpc/handler.go
@@ -131,7 +131,7 @@ func (h *handler) handleBatch(msgs []*jsonrpcMessage) {
 }
 
 // handleMsg handles a single message.
-func (h *handler) handleMsg(msg *jsonrpcMessage) {
+func (h *handler) handleMsg(msg *jsonrpcMessage, done chan(bool)) {
 	if ok := h.handleImmediate(msg); ok {
 		return
 	}
@@ -144,6 +144,9 @@ func (h *handler) handleMsg(msg *jsonrpcMessage) {
 		for _, n := range cp.notifiers {
 			n.activate()
 		}
+		if done != nil{
+			done <- true
+		}
 	})
 }
 
diff --git a/rpc/server.go b/rpc/server.go
index 5a92847f2..75b564743 100644
--- a/rpc/server.go
+++ b/rpc/server.go
@@ -111,7 +111,7 @@ func (s *Server) serveSingleRequest(ctx context.Context, codec ServerCodec) {
 	if batch {
 		h.handleBatch(reqs)
 	} else {
-		h.handleMsg(reqs[0])
+		h.handleMsg(reqs[0], nil)
 	}
 }
 
diff --git a/tests/testdata b/tests/testdata
index c02a2a17c..6b85703b5 160000
```

**Comment by fjl** [comment]
Thank you for the investigation! I'll fix it.

**Comment by fjl** [comment]
@holiman please try your example again, somehow it doesn't work for me. I said it's fixed because the unit test I added passes now but maybe it's different with unix sockets and real-life `nc`.

**Comment by fjl** [comment]
I'm still thinking about adding the parse errors back in.

**Comment by holiman** [comment]
Yup, the fix works for me. If you don't add parse errors back in, I'd suggest to instead bmp the error message from `DEBUG` to `INFO`. 
```
DEBUG[01-25|10:03:21.254] RPC connection read error                err="invalid character 'x' looking for beginning of value"
```
Otherwise, people will just have a hard time debugging their cross-platform woes, and they'll bug us about it.

**Comment by fjl** [comment]
rebased and parse errors are back

**Comment by shazow** [comment]
In case it's interesting to anyone involved in this, I also wrote a bidirectional jsonrpc2 implementation in Vipnode last August:

- https://godoc.org/github.com/vipnode/vipnode/jsonrpc2
- Example: https://github.com/vipnode/vipnode/blob/70acaee90fa7b0198a90d19637cdaac6999bde86/jsonrpc2/bidirpc_test.go

Let me know if there's desire to collaborate in the future.

**Pull request merged**